This patch introduces the functionality to preview a scrollback buffer.
Should be patched on top of (1) st-externalpipe, (2) st-externalpipe-eternal,
and (3) st-scrollback.
diff --git a/config.def.h b/config.def.h
index 051776d..b55b60a 100644
--- a/config.def.h
+++ b/config.def.h
@@ -246,5 +246,6 @@ static MouseShortcut mshortcuts[] = {
 static Shortcut shortcuts[] = {
 	/* mask                 keysym          function        argument */
+	{ TERMMOD,              XK_H,           openscrollback, {.s =  "/tmp/st-stdout"} },
 	{ XK_ANY_MOD,           XK_Break,       sendbreak,      {.i =  0} },
 	{ ControlMask,          XK_Print,       toggleprinter,  {.i =  0} },
 	{ ShiftMask,            XK_Print,       printscreen,    {.i =  0} },
diff --git a/st.c b/st.c
index 0600342..04bbae0 100644
--- a/st.c
+++ b/st.c
@@ -736,6 +736,52 @@ execsh(char *cmd, char **args)
 	_exit(1);
 }
 
+void
+kexecsh(const Arg * a)
+{
+	char buf[strlen(a->s) + 2];
+	snprintf(buf, sizeof buf, "%s%s", a->s, "\n");
+	ttywrite(buf, sizeof buf, 0);
+}
+
+void
+openscrollback(const Arg * a)
+{
+	char *catcommand, *lesscommand;
+	int catcommandsize, lesscommandsize;
+	int tmpfilesize = strlen(a->s);
+	char **writetotmp;
+	Arg pipe;
+	Arg arg;
+
+	/* Compose cat command */
+	catcommandsize = 6 + tmpfilesize + 1;
+	catcommand = (char *)malloc(catcommandsize);
+	strcpy(catcommand, "cat > ");
+	strcat(catcommand, a->s);
+
+	writetotmp = (char *[5]){
+		"/bin/sh",
+		"-c",
+		catcommand,
+		"externalpipe",
+		NULL
+	};
+	pipe.v = writetotmp;
+	/* Write scrollback buffer to tmp file */
+	externalpipe(&pipe);
+
+	/* Compose less command */
+	lesscommandsize = 5 + strlen(a->s) + 1;
+	lesscommand = (char *)malloc(lesscommandsize);
+	strcpy(lesscommand, "less ");
+	strcat(lesscommand, a->s);
+
+	/* Execute less */
+	arg.s = lesscommand;
+	kexecsh(&arg);
+}
+
 void
 sigchld(int a)
 {
diff --git a/st.h b/st.h
index 6fd0e40..742ff65 100644
--- a/st.h
+++ b/st.h
@@ -84,6 +84,8 @@ void draw(void);
 
 void kscrolldown(const Arg *);
 void kscrollup(const Arg *);
+void kexecsh(const Arg *);
+void openscrollback(const Arg *);
 void externalpipe(const Arg *);
 void printscreen(const Arg *);
 void printsel(const Arg *);
